### ISSUE TITLE
关于扫描方向实现问题

### ISSUE THREAD

**Issue by SmallLeishen** [bywlzts/CWNet-Causal-Wavelet-Network#13]
for layer in range(self.hnum):
            self.hhmamba.append(nn.ModuleList([
                SS2D6(d_model=dims, dropout=0, d_state=d_state, scan_type='lh'),
                PreNorm(dims, FeedForward(dim=dims))
            ]))
这是位于CWNet.py文件下，ProcessBlock类作为CWNet的核心操作，其初始化代码中，并没有创建不同扫描方向的mamba呀，因为scan_type字段不变，我发现，即使我改变scan_type的值，在SS2D6类中并没有任何地方使用到其值，反而是在SS2D类中使用了该字段，这是怎么回事，这里为何与论文中实现不一样呢？SS2D6能够实现不同方向扫描吗？他和SS2D什么关系，我看代码中SS2D6并未调用SS2D，到底是如何实现不同扫描方向的？
以下为您的代码中SS2D6的实现：
class SS2D6(nn.Module):
    def __init__(
            self,
            d_model,
            d_state=16,
            d_conv=3,
            expand=2,
            dt_rank="auto",
            dt_min=0.001,
            dt_max=0.1,
            dt_init="random",
            dt_scale=1.0,
            dt_init_floor=1e-4,
            dropout=0.,
            conv_bias=True,
            bias=False,
            device=None,
            dtype=None,
            scan_type='hl',
            **kwargs,
    ):
        factory_kwargs = {"device": device, "dtype": dtype}
        super().__init__()
        self.d_model = d_model
        self.d_state = d_state
        # self.d_state = math.ceil(self.d_model / 6) if d_state == "auto" else d_model # 20240109
        self.d_conv = d_conv
        self.expand = expand
        self.d_inner = int(self.expand * self.d_model)
        self.dt_rank = math.ceil(self.d_model / 16) if dt_rank == "auto" else dt_rank
        self.scantype = scan_type
        self.in_proj = nn.Linear(self.d_model, self.d_inner * 2, bias=bias, **factory_kwargs)
        self.conv2d = nn.Conv2d(
            in_channels=self.d_inner,
            out_channels=self.d_inner,
            groups=self.d_inner,
            bias=conv_bias,
            kernel_size=d_conv,
            padding=(d_conv - 1) // 2,
            **factory_kwargs,
        )
        self.act = nn.SiLU()

        self.x_proj = (
            nn.Linear(self.d_inner, (self.dt_rank + self.d_state * 2), bias=False, **factory_kwargs),
            nn.Linear(self.d_inner, (self.dt_rank + self.d_state * 2), bias=False, **factory_kwargs),
            nn.Linear(self.d_inner, (self.dt_rank + self.d_state * 2), bias=False, **factory_kwargs),
            nn.Linear(self.d_inner, (self.dt_rank + self.d_state * 2), bias=False, **factory_kwargs),
            nn.Linear(self.d_inner, (self.dt_rank + self.d_state * 2), bias=False, **factory_kwargs),
            nn.Linear(self.d_inner, (self.dt_rank + self.d_state * 2), bias=False, **factory_kwargs)
        )
        self.x_proj_weight = nn.Parameter(torch.stack([t.weight for t in self.x_proj], dim=0)) # (K=4, N, inner)

        del self.x_proj

        self.dt_projs = (
            self.dt_init(self.dt_rank, self.d_inner, dt_scale, dt_init, dt_min, dt_max, dt_init_floor, **factory_kwargs),
            self.dt_init(self.dt_rank, self.d_inner, dt_scale, dt_init, dt_min, dt_max, dt_init_floor, **factory_kwargs),
            self.dt_init(self.dt_rank, self.d_inner, dt_scale, dt_init, dt_min, dt_max, dt_init_floor, **factory_kwargs),
            self.dt_init(self.dt_rank, self.d_inner, dt_scale, dt_init, dt_min, dt_max, dt_init_floor, **factory_kwargs),
            self.dt_init(self.dt_rank, self.d_inner, dt_scale, dt_init, dt_min, dt_max, dt_init_floor, **factory_kwargs),
            self.dt_init(self.dt_rank, self.d_inner, dt_scale, dt_init, dt_min, dt_max, dt_init_floor, **factory_kwargs)
        )
        self.dt_projs_weight = nn.Parameter(torch.stack([t.weight for t in self.dt_projs], dim=0)) # (K=4, inner, rank)
        self.dt_projs_bias = nn.Parameter(torch.stack([t.bias for t in self.dt_projs], dim=0)) # (K=4, inner)
        del self.dt_projs

        self.A_logs = self.A_log_init(self.d_state, self.d_inner, copies=6, merge=True) # (K=4, D, N)
        self.Ds = self.D_init(self.d_inner, copies=6, merge=True) # (K=4, D, N)

        self.selective_scan = selective_scan_fn
        self.forward_core = self.forward_corev0
        self.diagonal_trans = self.diagonal_trans
        self.reverse_diagonal_trans = self.reverse_diagonal_trans

        self.out_norm = nn.LayerNorm(self.d_inner)
        self.out_proj = nn.Linear(self.d_inner, self.d_model, bias=bias, **factory_kwargs)
        self.dropout = nn.Dropout(dropout) if dropout > 0. else None

    @staticmethod
    def dt_init(dt_rank, d_inner, dt_scale=1.0, dt_init="random", dt_min=0.001, dt_max=0.1, dt_init_floor=1e-4,
                **factory_kwargs):
        dt_proj = nn.Linear(dt_rank, d_inner, bias=True, **factory_kwargs)

        # Initialize special dt projection to preserve variance at initialization
        dt_init_std = dt_rank ** -0.5 * dt_scale
        if dt_init == "constant":
            nn.init.constant_(dt_proj.weight, dt_init_std)
        elif dt_init == "random":
            nn.init.uniform_(dt_proj.weight, -dt_init_std, dt_init_std)
        else:
            raise NotImplementedError

        # Initialize dt bias so that F.softplus(dt_bias) is between dt_min and dt_max
        dt = torch.exp(
            torch.rand(d_inner, **factory_kwargs) * (math.log(dt_max) - math.log(dt_min))
            + math.log(dt_min)
        ).clamp(min=dt_init_floor)
        # Inverse of softplus: https://github.com/pytorch/pytorch/issues/72759
        inv_dt = dt + torch.log(-torch.expm1(-dt))
        with torch.no_grad():
            dt_proj.bias.copy_(inv_dt)
        # Our initialization would set all Linear.bias to zero, need to mark this one as _no_reinit
        dt_proj.bias._no_reinit = True

        return dt_proj

    @staticmethod
    def A_log_init(d_state, d_inner, copies=1, device=None, merge=True):
        # S4D real initialization
        A = repeat(
            torch.arange(1, d_state + 1, dtype=torch.float32, device=device),
            "n -> d n",
            d=d_inner,
        ).contiguous()
        A_log = torch.log(A)  # Keep A_log in fp32
        if copies > 1:
            A_log = repeat(A_log, "d n -> r d n", r=copies)
            if merge:
                A_log = A_log.flatten(0, 1)
        A_log = nn.Parameter(A_log)
        A_log._no_weight_decay = True
        return A_log

    @staticmethod
    def D_init(d_inner, copies=1, device=None, merge=True):
        # D "skip" parameter
        D = torch.ones(d_inner, device=device)
        if copies > 1:
            D = repeat(D, "n1 -> r n1", r=copies)
            if merge:
                D = D.flatten(0, 1)
        D = nn.Parameter(D)  # Keep in fp32
        D._no_weight_decay = True
        return D

    def forward_corev0(self, x: torch.Tensor):
        self.selective_scan = selective_scan_fn

        B, C, H, W = x.shape
        L = H * W
        K = 6

        x_hwwh = torch.stack([x.view(B, -1, L), torch.transpose(x, dim0=2, dim1=3).contiguous().view(B, -1, L)],
                             dim=1).view(B, 2, -1, L)
        xs = torch.cat([x_hwwh, torch.flip(x_hwwh, dims=[-1])], dim=1)  # (b, k, d, l)

        x_invhh = torch.flip(x, dims=[-1])
        x_hh = torch.stack([x.view(B, -1, L), x_invhh.view(B, -1, L)], dim=1).view(B, 2, -1, L)
        x_hh, diag_indices = self.diagonal_trans(x_hh, H, W)

        xs = torch.cat([xs, x_hh], dim=1)  # (b, k, d, l)

        x_dbl = torch.einsum("b k d l, k c d -> b k c l", xs.view(B, K, -1, L), self.x_proj_weight)
        # x_dbl = x_dbl + self.x_proj_bias.view(1, K, -1, 1)
        dts, Bs, Cs = torch.split(x_dbl, [self.dt_rank, self.d_state, self.d_state], dim=2)
        dts = torch.einsum("b k r l, k d r -> b k d l", dts.view(B, K, -1, L), self.dt_projs_weight)
        # dts = dts + self.dt_projs_bias.view(1, K, -1, 1)

        xs = xs.float().view(B, -1, L)  # (b, k * d, l)
        dts = dts.contiguous().float().view(B, -1, L)  # (b, k * d, l)
        Bs = Bs.float().view(B, K, -1, L)  # (b, k, d_state, l)
        Cs = Cs.float().view(B, K, -1, L)  # (b, k, d_state, l)
        Ds = self.Ds.float().view(-1)  # (k * d)
        As = -torch.exp(self.A_logs.float()).view(-1, self.d_state)  # (k * d, d_state)
        dt_projs_bias = self.dt_projs_bias.float().view(-1)  # (k * d)

        out_y = self.selective_scan(
            xs, dts,
            As, Bs, Cs, Ds, z=None,
            delta_bias=dt_projs_bias,
            delta_softplus=True,
            return_last_state=False,
        ).view(B, K, -1, L)
        assert out_y.dtype == torch.float

        inv_y = torch.flip(out_y[:, 2:4], dims=[-1]).view(B, 2, -1, L)
        wh_y = torch.transpose(out_y[:, 1].view(B, -1, W, H), dim0=2, dim1=3).contiguous().view(B, -1, L)
        invwh_y = torch.transpose(inv_y[:, 1].view(B, -1, W, H), dim0=2, dim1=3).contiguous().view(B, -1, L)

        transhh_y = self.reverse_diagonal_trans(out_y[:, 4:6], diag_indices)
        invhh_y = torch.flip(transhh_y[:, 1], dims=[-1])

        return out_y[:, 0], inv_y[:, 0], wh_y, invwh_y, transhh_y[:, 0], invhh_y

    def diagonal_trans(self, x: torch.Tensor, H, W):
        B, K, _, L = x.shape
        assert L == H * W, "L==H*W"
        idx = torch.arange(H * W, device=x.device).reshape(H, W)
        i_idx = torch.arange(H, device=x.device).reshape(-1, 1).expand(H, W)
        j_idx = torch.arange(W, device=x.device).reshape(1, -1).expand(H, W)
        diag_mask = (i_idx + j_idx)
        sorted_idx = torch.argsort(diag_mask.view(-1))
        diag_indices = torch.index_select(idx.view(-1), 0, sorted_idx)
        x_flat = x.view(B, K, -1, H * W)
        diag_flat = torch.index_select(x_flat, dim=3, index=diag_indices)
        return diag_flat, diag_indices

    def reverse_diagonal_trans(self, x: torch.Tensor, diag_indices):
        B, K, _, L = x.shape
        x_flat = x.view(B, K, -1, L)
        reverse_indices = torch.argsort(diag_indices)
        original_flat = torch.index_select(x_flat, dim=3, index=reverse_indices)
        return original_flat

    def forward(self, x: torch.Tensor, **kwargs):
        B, H, W, C = x.shape
        xz = self.in_proj(x)
        x, z = xz.chunk(2, dim=-1)  # (b, h, w, d)

        x = x.permute(0, 3, 1, 2).contiguous()
        x = self.act(self.conv2d(x))  # (b, d, h, w)
        y1, y2, y3, y4, y5, y6 = self.forward_core(x)
        assert y1.dtype == torch.float32
        y = y1 + y2 + y3 + y4 + y5 + y6
        y = torch.transpose(y, dim0=1, dim1=2).contiguous().view(B, H, W, -1)
        y = self.out_norm(y)
        y = y * F.silu(z)
        out = self.out_proj(y)
        if self.dropout is not None:
            out = self.dropout(out)

        return out

**Comment by bywlzts** [comment]
您好，实际上scan_type这个参数并没有实际使用，这里没有给删掉，这是初始化类遗留下的一个无效参数，具体地修改是在对一维序列的变换上，具体是x_invhh = torch.flip(x, dims=[-1])
    x_hh = torch.stack([x.view(B, -1, L), x_invhh.view(B, -1, L)], dim=1).view(B, 2, -1, L)
    x_hh, diag_indices = self.diagonal_trans(x_hh, H, W)，这部分

**Comment by SmallLeishen** [comment]
for layer in range(self.hnum):
        self.hhmamba.append(nn.ModuleList([
            SS2D6(d_model=dims, dropout=0, d_state=d_state, scan_type='lh'),
            PreNorm(dims, FeedForward(dim=dims))
        ]))
您好！，因为SS2D6的代码似乎并没有实现不同模式的扫描，如果我把上买的SS2D6，替换成SS2D。如下：
self.hfmamba = nn.ModuleList([
            SS2D(d_model=self.one, d_state=16, dropout=0.1, scan_type='lh'),
            PreNorm(self.one, FeedForward(dim=self.one))
        ])
那么是否可以实现垂直方向的扫描（即主要捕捉那些横向边缘的细节信息），然后针对不同类型的输入，改变scan_type为'hl'或者'hh'就可以实现特定方式的扫描？我的理解对吗？

**Comment by SmallLeishen** [comment]
对于您的代码，我有一些理解想请您帮我矫正以下：
小波分解可以得到四个不同的频带，其中hl表示水平高频，即该频带捕获了大量垂直的线条（垂直线条在水平方向上像素变化大），然后对于该频带的特征适合用横向扫描的mamba来处理，因为沿着横向的方向，全部都是高频信息，这样可以提取到更有用的特征。对吗？

**Comment by bywlzts** [comment]
> for layer in range(self.hnum): self.hhmamba.append(nn.ModuleList([ SS2D6(d_model=dims, dropout=0, d_state=d_state, scan_type='lh'), PreNorm(dims, FeedForward(dim=dims)) ])) 您好！，因为SS2D6的代码似乎并没有实现不同模式的扫描，如果我把上买的SS2D6，替换成SS2D。如下： self.hfmamba = nn.ModuleList([ SS2D(d_model=self.one, d_state=16, dropout=0.1, scan_type='lh'), PreNorm(self.one, FeedForward(dim=self.one)) ]) 那么是否可以实现垂直方向的扫描（即主要捕捉那些横向边缘的细节信息），然后针对不同类型的输入，改变scan_type为'hl'或者'hh'就可以实现特定方式的扫描？我的理解对吗？

不是的，实际的实现我并没有用scan_type这个参数，是固定的6个扫描方向，分别对应这六个返回值：return out_y[:, 0], inv_y[:, 0], wh_y, invwh_y, transhh_y[:, 0], invhh_y，您可以追溯代码看一下。第二个问题，关于扫描方向的理解您是对的。
